pkg/clients/subnetwork/subnetwork.go

```diff
@@ -42,28 +79,71 @@ func GenerateSubnetwork(s v1alpha3.SubnetworkParameters) *compute.Subnetwork {
-	s.Region = in.Region
-	for _, val := range in.SecondaryIpRanges {
-		obj := &v1alpha3.GCPSubnetworkSecondaryRange{
-			IPCidrRange: val.IpCidrRange,
-			RangeName:   val.RangeName,
+// GenerateSubnetworkObservation creates a SubnetworkObservation object using *googlecompute.Subnetwork.
+func GenerateSubnetworkObservation(in compute.Subnetwork) v1beta1.SubnetworkObservation {
+	return v1beta1.SubnetworkObservation{
+		CreationTimestamp: in.CreationTimestamp,
+		Fingerprint:       in.Fingerprint,
+		GatewayAddress:    in.GatewayAddress,
+		ID:                in.Id,
+		SelfLink:          in.SelfLink,
+	}
+}
+
+// LateInitializeSpec fills unassigned fields with the values in compute.Subnetwork object.
+func LateInitializeSpec(spec *v1beta1.SubnetworkParameters, in compute.Subnetwork) {
+	if spec.IPCidrRange == "" {
+		spec.IPCidrRange = in.IpCidrRange
+	}
+
+	if spec.Region == "" {
+		spec.Region = in.Region
+	}
+
+	spec.Network = gcp.LateInitializeString(spec.Network, in.Network)
+	spec.Description = gcp.LateInitializeString(spec.Description, in.Description)
+	spec.EnableFlowLogs = gcp.LateInitializeBool(spec.EnableFlowLogs, in.EnableFlowLogs)
+	spec.PrivateIPGoogleAccess = gcp.LateInitializeBool(spec.PrivateIPGoogleAccess, in.PrivateIpGoogleAccess)
+	if len(in.SecondaryIpRanges) != 0 && len(spec.SecondaryIPRanges) == 0 {
+		spec.SecondaryIPRanges = make([]*v1beta1.SubnetworkSecondaryRange, len(in.SecondaryIpRanges))
+		for i, r := range in.SecondaryIpRanges {
+			spec.SecondaryIPRanges[i] = &v1beta1.SubnetworkSecondaryRange{
+				IPCidrRange: r.IpCidrRange,
+				RangeName:   r.RangeName,
+			}
 		}
-		s.SecondaryIPRanges = append(s.SecondaryIPRanges, obj)
 	}
-	s.CreationTimestamp = in.CreationTimestamp
-	s.GatewayAddress = in.GatewayAddress
-	s.ID = in.Id
-	s.Kind = in.Kind
-	s.SelfLink = in.SelfLink
-	return s
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(name string, in *v1beta1.SubnetworkParameters, observed *compute.Subnetwork) (upToDate bool, privateAccess bool, err error) {
+	generated, err := copystructure.Copy(observed)
+	if err != nil {
+		return true, false, errors.Wrap(err, errCheckUpToDate)
+	}
+	desired, ok := generated.(*compute.Subnetwork)
+	if !ok {
+		return true, false, errors.New(errCheckUpToDate)
+	}
+	GenerateSubnetwork(name, *in, desired)
+	if !cmp.Equal(desired.PrivateIpGoogleAccess, observed.PrivateIpGoogleAccess) {
+		return false, true, nil
+	}
+	return cmp.Equal(desired, observed, cmpopts.EquateEmpty(), cmpopts.SortSlices(equateGCPSecondaryRange), equatePrefixedStrings(v1beta1.URIPrefix, v1beta1.URIRegionPrefix)), false, nil
+}
+
+func equatePrefixedStrings(prefixes ...string) cmp.Option {
+	return cmp.Comparer(func(a, b string) bool {
+		if a == b {
+			return true
+		}
+
+		for _, prefix := range prefixes {
+			if strings.TrimPrefix(a, prefix) == strings.TrimPrefix(b, prefix) {
+				return true
+			}
+		}
+		return false
+	})
```

Comment:
This was added to fix comparison of the `region` and `network` fields, which GCP returns with URI prefixes. @negz would you support me adding this to `cmp` funcs in `crossplane-runtime`?